chaos/net-rpc-client.go

```diff
@@ -26,135 +25,104 @@ import (
-	return rpcLocalStack
-}
-
-// dialRPCClient tries to establish a connection to the server in a safe manner
-func (rpcClient *ReconnectRPCClient) dialRPCClient() (*rpc.Client, error) {
-	rpcClient.mu.Lock()
-	defer rpcClient.mu.Unlock()
-	// After acquiring lock, check whether another thread may not have already dialed and established connection
-	if rpcClient.rpcPrivate != nil {
-		return rpcClient.rpcPrivate, nil
-	}
-	rpc, err := rpc.DialHTTPPath("tcp", rpcClient.node, rpcClient.rpcPath)
-	if err != nil {
-		return nil, err
-	} else if rpc == nil {
-		return nil, errors.New("No valid RPC Client created after dial")
+// Close closes the underlying socket file descriptor.
+func (rpcClient *ReconnectRPCClient) Close() error {
+	rpcClient.mutex.Lock()
+	defer rpcClient.mutex.Unlock()
+	// If rpc client has not connected yet there is nothing to close.
+	if rpcClient.rpc == nil {
+		return nil
 	}
-	rpcClient.rpcPrivate = rpc
-	return rpcClient.rpcPrivate, nil
+	// Reset rpcClient.rpc to allow for subsequent calls to use a new
+	// (socket) connection.
+	clnt := rpcClient.rpc
+	rpcClient.rpc = nil
+	return clnt.Close()
 }
 
 // Call makes a RPC call to the remote endpoint using the default codec, namely encoding/gob.
 func (rpcClient *ReconnectRPCClient) Call(serviceMethod string, args interface{}, reply interface{}) error {
-	// Make a copy below so that we can safely (continue to) work with the rpc.Client.
-	// Even in the case the two threads would simultaneously find that the connection is not initialised,
-	// they would both attempt to dial and only one of them would succeed in doing so.
-	rpcLocalStack := rpcClient.getRPCClient()
-
+	rpcClient.mutex.Lock()
+	defer rpcClient.mutex.Unlock()
 	// If the rpc.Client is nil, we attempt to (re)connect with the remote endpoint.
-	if rpcLocalStack == nil {
-		var err error
-		rpcLocalStack, err = rpcClient.dialRPCClient()
+	if rpcClient.rpc == nil {
+		clnt, err := rpc.DialHTTPPath("tcp", rpcClient.addr, rpcClient.endpoint)
 		if err != nil {
 			return err
 		}
+		rpcClient.rpc = clnt
 	}
 
 	// If the RPC fails due to a network-related error, then we reset
 	// rpc.Client for a subsequent reconnect.
-	err := rpcLocalStack.Call(serviceMethod, args, reply)
-	if err != nil {
-		if err.Error() == rpc.ErrShutdown.Error() {
-			// Reset rpcClient.rpc to nil to trigger a reconnect in future
-			// and close the underlying connection.
-			rpcClient.clearRPCClient()
-
-			// Close the underlying connection.
-			rpcLocalStack.Close()
-
-			// Set rpc error as rpc.ErrShutdown type.
-			err = rpc.ErrShutdown
-		}
+	err := rpcClient.rpc.Call(serviceMethod, args, reply)
+	if err == rpc.ErrShutdown {
+		rpcClient.rpc.Close()
+		rpcClient.rpc = nil
 	}
 	return err
 }
 
 func (rpcClient *ReconnectRPCClient) RLock(args dsync.LockArgs) (status bool, err error) {
 	err = rpcClient.Call("Dsync.RLock", &args, &status)
+	if err == rpc.ErrShutdown {
+		err = rpcClient.Call("Dsync.RLock", &args, &status)
```

Comment:
There is no need to retry at a NetLocker implementation layer. The lock retry is done at DRWMutex.Rlock layer.